Telcodocs 2500: GNSS (GPS) as the Primary System Clock Time Source with NTP Fallback (#103178)
openshift/openshift-docs: modules/nw-ptp-configuring-gnss-to-ntp-failover-sno.adoc

This procedure configures a T-GM (Telecom Grandmaster) clock on Single Node OpenShift (SNO) that uses an Intel E810 Westport Channel NIC as the PTP grandmaster clock with GNSS to NTP failover capabilities.

.Prerequisites

* For T-GM clocks in production environments, install an Intel E810 Westport Channel NIC in the bare-metal SNO host.

* Install the OpenShift CLI (`oc`).

* Log in as a user with `cluster-admin` privileges.

* Install the PTP Operator.

.Procedure

. Verify the PTP Operator installation by running the following command:
+
[source,terminal]
----
$ oc get pods -n openshift-ptp -o wide
----
+
The output is similar to the following listing the PTP Operator pod and the single `linuxptp-daemon` pod:
+
[source,terminal]
----
NAME                            READY   STATUS    RESTARTS   AGE   IP              NODE                   NOMINATED NODE   READINESS GATES
linuxptp-daemon-xz8km           2/2     Running   0          15m   192.168.1.50    mysno-sno.demo.lab     <none>           <none>
ptp-operator-75c77dbf86-xm9kl   1/1     Running   0          20m   10.129.0.45     mysno-sno.demo.lab     <none>           <none>
----
+

> * `ptp-operator-*`: The PTP Operator pod (one instance in the cluster)

Reviewer: ```suggestion
* `ptp-operator-*`: The PTP Operator pod (one instance in the cluster).
```

Added stop to be consistent with the next item.

* `linuxptp-daemon-*`: The linuxptp daemon pod. On SNO, there is only one daemon pod running on the master node. The daemon pod should show `2/2` in the READY column, indicating both containers (`linuxptp-daemon-container` and `kube-rbac-proxy`) are running.

. Check which network interfaces support hardware timestamping by running the following command:
+
[source,terminal]
----
$ oc get NodePtpDevice -n openshift-ptp -o yaml
----
+

> The output is similar to the following showing the NodePtpDevice resource for the SNO node with PTP-capable network interfaces:

Reviewer: ```suggestion
The output is similar to the following one, showing the NodePtpDevice resource for the SNO node with PTP-capable network interfaces:
```

Added for clarity. By "following", we mean the following output?

. Check which network interfaces support hardware timestamping by running the following command:
+
[source,terminal]
----
$ oc get NodePtpDevice -n openshift-ptp -o yaml
----
+
The output is similar to the following one, showing the NodePtpDevice resource for the SNO node with PTP-capable network interfaces:
+
[source,yaml]
----
apiVersion: v1
items:
- apiVersion: ptp.openshift.io/v1
  kind: NodePtpDevice
  metadata:
    name: mysno-sno.demo.lab
    namespace: openshift-ptp
  spec: {}
  status:
    devices:
    - name: ens7f0
      hwConfig:
        phcIndex: 0
    - name: ens7f1
      hwConfig:
        phcIndex: 1
kind: List
metadata:
  resourceVersion: ""
----
+
In this example output:
+
* `ens7f0` and `ens7f1` are PTP-capable interfaces (Intel E810 NIC ports).
* `phcIndex` indicates the PTP Hardware Clock number (maps to `/dev/ptp0`, `/dev/ptp1`, etc.)
+
[NOTE]
====
On SNO clusters, you will see only one NodePtpDevice resource for the single master node.
====

> . The PTP profile uses node labels for matching. Check your Machine Config Pool (MCP) to verify the master MCP. Run the following command:

Reviewer: ```suggestion
. The PTP profile uses node labels for matching. Check your Machine Config Pool (MCP) to verify the master MCP by running the following command:
```

Little style suggestion.

. Verify the PTP Operator installation by running the following command:
+
[source,terminal]
----
$ oc get pods -n openshift-ptp -o wide
----
+
The output is similar to the following listing the PTP Operator pod and the single `linuxptp-daemon` pod:
+
[source,terminal]
----
NAME                            READY   STATUS    RESTARTS   AGE   IP              NODE                   NOMINATED NODE   READINESS GATES
linuxptp-daemon-xz8km           2/2     Running   0          15m   192.168.1.50    mysno-sno.demo.lab     <none>           <none>
ptp-operator-75c77dbf86-xm9kl   1/1     Running   0          20m   10.129.0.45     mysno-sno.demo.lab     <none>           <none>
----
+
* `ptp-operator-*`: The PTP Operator pod (one instance in the cluster).
* `linuxptp-daemon-*`: The linuxptp daemon pod. On {sno}, there is only one daemon pod running on the master node. The daemon pod should show `2/2` in the READY column, indicating both containers (`linuxptp-daemon-container` and `kube-rbac-proxy`) are running.

. Check which network interfaces support hardware timestamping by running the following command:
+
[source,terminal]
----
$ oc get NodePtpDevice -n openshift-ptp -o yaml
----
+

> The output is similar to the following one, showing the NodePtpDevice resource for the {sno} node with PTP-capable network interfaces:

Reviewer 2: ```suggestion
The output is similar to the following one, showing the `NodePtpDevice` resource for the {sno} node with PTP-capable network interfaces:
```

      delay_mechanism E2E
      time_stamping hardware
      tsproc_mode filter
      delay_filter moving_median
      delay_filter_length 10
      egressLatency 0
      ingressLatency 0
      boundary_clock_jbod 0
      #
      # Clock description
      #
      productDescription ;;
      revisionData ;;
      manufacturerIdentity 00:00:00
      userDescription ;
      timeSource 0x20
    ptpClockThreshold:
      holdOverTimeout: 5
      maxOffsetThreshold: 100
      minOffsetThreshold: -100
  recommend:
  - profile: "grandmaster"
    priority: 4
    match:
    - nodeLabel: node-role.kubernetes.io/master
----
+
[IMPORTANT]
====
Replace the example interface names (`ens7f0`, `ens7f1`) with your actual E810 NIC interface names found in step 2. Common E810 interface naming patterns include `ens7f0`, `ens8f0`, `eth0`, `enp2s0f0`, and so on. The exact name depends on your system BIOS settings and Linux network device naming conventions. Also, replace `/dev/ttyGNSS_1700_0` with your actual GNSS serial port device path. The `nodeLabel` is set to `node-role.kubernetes.io/master` to target the {sno} master node which serves all roles.
====
+
Key Configuration Components:
+

> **PTP4L Options:**

Reviewer 2: same question here as on ln 396 of non-SNO version:
Is the indentation in the preview displaying as intended?
